service/matching/tasklist/isolation_balancer.go

```diff
@@ -0,0 +1,422 @@
+			toRemove = append(toRemove, group)
+		}
+	}
+	// partition id to set of groups
+	partitionGroups := make(map[int]map[string]any)
+	for id, partition := range partitions {
+		set := make(map[string]any)
+		for _, group := range partition.IsolationGroups {
+			set[group] = true
+		}
+		partitionGroups[id] = set
+	}
+	removePartitions(toRemove, groupSizePerPartition, partitionGroups)
+	addPartitions(toAdd, groupSizePerPartition, partitionGroups)
+	minimumGroups := i.getMinGroupsPerPartition(assignableGroups)
+	movedGroupBetweenPartitions := ensureMinimumGroupsPerPartition(minimumGroups, groupSizePerPartition, partitionGroups)
+	result := make(map[int]*types.TaskListPartition)
+	for id, groups := range partitionGroups {
+		if len(groups) == 0 {
+			result[id] = &types.TaskListPartition{}
+		} else {
+			asSlice := maps.Keys(groups)
+			// Sort for the sake of stability
+			slices.Sort(asSlice)
+			result[id] = &types.TaskListPartition{IsolationGroups: asSlice}
+		}
+	}
+	return result, movedGroupBetweenPartitions || len(toRemove) > 0 || len(toAdd) > 0
+}
+
+func (i *isolationBalancer) getMinPartitionsPerGroup(writePartitionCount, assignableGroups int) int {
+	if assignableGroups == 0 {
+		return 0
+	}
+	return int(math.Ceil(float64(writePartitionCount*i.getMinGroupsPerPartition(assignableGroups)) / float64(assignableGroups)))
+}
+
+func (i *isolationBalancer) getMinGroupsPerPartition(assignableGroups int) int {
+	return min(i.config.IsolationGroupsPerPartition(), assignableGroups)
+}
+
+func removePartitions(toRemove []string, groupSizePerPartition map[string]float64, groupsByPartitionID map[int]map[string]any) {
+	// Use a greedy approach to remove the specified groups (which may include duplicates) from the largest partition
+	// Similarly to adding partitions we sort the input by size first
+	slices.SortFunc(toRemove, byValueDescending(groupSizePerPartition))
+	for _, group := range toRemove {
+		partitionID, ok := findMaxWithGroup(group, groupSizePerPartition, groupsByPartitionID)
+		if ok {
+			delete(groupsByPartitionID[partitionID], group)
+		}
+	}
+}
+
+func addPartitions(toAdd []string, groupSizePerPartition map[string]float64, groupsByPartitionID map[int]map[string]any) {
+	// Use greedy number partitioning to add the group to the smallest partition where it isn't already present. This
+	// is an approximate solution. Sorting the input by size in a descending manner should result in a more even
+	// distribution.
+	slices.SortFunc(toAdd, byValueDescending(groupSizePerPartition))
+	for _, group := range toAdd {
+		partitionID, ok := findMinWithoutGroup(group, groupSizePerPartition, groupsByPartitionID)
+		if ok {
+			groupsByPartitionID[partitionID][group] = true
+		}
+	}
+}
+
+func ensureMinimumGroupsPerPartition(minGroups int, groupSizePerPartition map[string]float64, groupsByPartitionID map[int]map[string]any) bool {
+	// Every partition needs to be assigned groups that have pollers, otherwise we're unable to process tasks on those
```

Comment:
I suppose a very general question which probably has been discussed already: I think the underlying assumption here is that a partition will always be able to handle the load from an isolation group, is that correct? Is there a limitation implied here on isolation-groups which may have more traffic than a partition can tolerate? Is this even a problem worth caring about? (sincere questions, I'm quite out of the loop on this one)

Reply:
(after discussing offline) I missed the core mechanic of the autoscaler interacting with this, so that makes sense. ty.